kimchi/src/alphas.rs

```diff
@@ -38,120 +39,114 @@ use std::{
 #[derive(Debug, Default, Serialize, Deserialize, Clone)]
-pub struct Builder {
+pub struct Alphas<F> {
     /// The next power of alpha to use
     /// the end result will be [1, alpha^next_power)    
     next_power: usize,
     /// The mapping between constraint types and powers of alpha
-    mapping: HashMap<ConstraintType, Range<usize>>,
+    mapping: HashMap<ArgumentType, Range<usize>>,
+    /// The powers of alpha: 1, alpha, alpha^2, etc.
+    /// If set to [Some], you can't register new constraints.
+    alphas: Option<Vec<F>>,
 }
 
-impl Builder {
-    /// Registers a new [ConstraintType],
+impl<F: Field> Alphas<F> {
+    /// Registers a new [ArgumentType],
     /// associating it with a number `powers` of powers of alpha.
-    /// The function returns an iterator of powers that haven't been used yet.
-    pub fn register(&mut self, ty: ConstraintType, powers: usize) -> Range<usize> {
+    pub fn register(&mut self, ty: ArgumentType, powers: usize) {
+        if self.alphas.is_some() {
+            panic!("you cannot register new constraints once initialized with a field element");
+        }
         let new_power = self.next_power + powers;
         let range = self.next_power..new_power;
-        if self.mapping.insert(ty, range.clone()).is_some() {
-            panic!("cannot re-register {:?}", ty);
-        }
-        self.next_power = new_power;
-        range
-    }
-}
-
-// ------------------------------------------
 
-/// This type can be constructed from [Builder] and a value `alpha`,
-/// via [Alphas::new].
-/// It will then pre-compute the necessary numbers of powers of alpha.
-/// You can retrieve powers of alpha by calling [Alphas::take_alphas],
-/// or the exponents without the actual powers of alphas via [Alphas::take_powers].
-pub struct Alphas<F> {
-    /// The powers of alpha: 1, alpha, alpha^2, etc.
-    alphas: Vec<F>,
-    /// The mapping from constraint type to a range of powers of alpha
-    mapping: HashMap<ConstraintType, Range<usize>>,
-}
-
-impl<F: Field> Alphas<F> {
-    /// Creates a new instance of [Alphas] via a [Builder] and value `alpha`.
-    pub fn new(alpha: F, powers: &Builder) -> Alphas<F> {
-        let mut last_power = F::one();
-        let mut alphas = Vec::with_capacity(powers.next_power);
-        alphas.push(F::one());
-        for _ in 0..(powers.next_power - 1) {
-            last_power *= alpha;
-            alphas.push(last_power);
-        }
+        // gates are a special case, as we reuse the same power of alpha
+        // across all of them (they're mutually exclusive)
+        let ty = if matches!(ty, ArgumentType::Gate(_)) {
+            // the zero gate is not used, so we default to it
+            ArgumentType::Gate(GateType::Zero)
+        } else {
+            ty
+        };
 
-        Alphas {
-            alphas,
-            mapping: powers.mapping.clone(),
+        if self.mapping.insert(ty, range).is_some() {
```

Comment:
I am a bit confused about the functionality of this check, but perhaps a short comment could solve it. In particular, it's about how the ranges increase, or when this `register()` method should be invoked. Before the new alphas merged, we used to have ranges that grew consecutively together with the `GateType`, up to around power 79 or so for `ChaChaFinal`. Then, after the alphas PR, I was thinking we were sticking to 21 powers or so (the largest number of constraints, in any available gate). And as long as the selector polynomials were mutually exclusive, that was fine. 
But here, it seems to suggest that one could potentially have "duplicate" entries of `GateTypes` with different ranges, right? I am assuming this won't ever be happening in the code, as this `register()` method will only be called once per `GateType`. But I would like to double check with you that this is indeed the case. If so, perhaps it is safer if the mapping checked only the `ty` part, regardless the range? Or is there a reason why the check is made on the full tuple `(ty, range)`?

Reply:
If I understand your question correctly, you're misreading this line: 

* this inserts `range` under the mapping for `ty` (`Gate`, `Permutation`, or `Lookup`). 
* Importantly, if this is a `Gate(_)`, the range will always be stored under `Gate(Zero)` due to the line above (so you CAN'T have disjoint ranges for different gates). 
* The result of inserting is either `None` (all good, we never registered that argument before), or the previous value `Some(_)` in which case we panic because there's already a range registered under that argument.

Reply:
Oooh I see! I think I was misunderstanding the possible values that `ty` could take! In particular, I had forgotten that it was only `Gate`, `Permutation` and `Lookup`, and I was assuming it took any `GateType` such as `Poseidon` or `ChaCha1`. Thanks, this clarifies everything.